lib/waterline/query/dql/tenant.js

```diff
@@ -0,0 +1,194 @@
+
+/**
+ * Tenant
+ *
+ * @param {Object} criteria
+ * @param {Object} options
+ * @param {Function} callback
+ * @return Deferred object if no callback
+ */
+
+module.exports = function(tenant, cb) {
+  var self = this;
+  // console.log('tenant', arguments);
+  var usage = utils.capitalize(this.identity) + '.tenant(tenantId,callback)';
+
+  if (arguments.length === 0) {
+    //
+    // Error
+    return usageError('Tenant must be passed.', usage, cb);
+  }
+
+  if (typeof tenant === 'function') {
+    // Tenant is required
+    // Error
+    return usageError('Tenant must be passed.', usage, cb);
+  }
+
+
+  // Change connection settings based on tenant
+  // console.log('collections', this.waterline.collections);
+  // console.log('adapterDictionary', this.waterline, this.adapterDictionary);
+
+  var baseCollectionIdentity = self.baseIdentity || self.identity;
+  var connectionName = this.adapterDictionary['identity'];
+  var baseConnectionName = connectionName;
+  if (baseCollectionIdentity != self.identity) {
+      // Is not the base Collection
+      // Get base collection
+      var baseCollection = self.waterline.collections[baseCollectionIdentity];
+    //   console.log('baseCollection', baseCollectionIdentity);
+      baseConnectionName = baseCollection.adapterDictionary['identity'];
+    //   console.log('baseConnectionName', baseConnectionName);
+  }
+  var connection = this.connections[connectionName];
+  // Change for Multi-Tenancy support
+  // console.log('connection', connection);
+  var config = connection.config;
+  // console.log('config', config);
+
+  if (config.isMultiTenant && typeof config.getTenant === "function") {
+
+    var configForTenant = config.configForTenant;
+
+    // Return Deferred or pass to adapter
+    if (typeof cb === 'function') {
+
+      var configDup = _.cloneDeep(config);
+      configDup.tenant = tenant; // Set Tenant
+      configForTenant.call(self, tenant, configDup, function(err, config) {
+        // console.log('configForTenant', err, config);
+        if (err) return cb(err);
+
+        // Tenant's config namespace
+        var tenantSuffix = "<tenant:" + tenant + ">";
+        var newConnectionIdentity = (baseConnectionName + tenantSuffix).toLowerCase();
+
+        function createTenantCollection() {
+
+            var newCollectionIdentity = (baseCollectionIdentity + tenantSuffix).toLowerCase();
+
+            // Take new config and register it's connection
+            // var cols = self.waterline._collections;
+            // console.log('collections', cols);
+            // var col = _.find(cols, function(col) {
+            //     return true;
+            // });
+            // var col = self.collection;
+            // console.log(col);
+            // var colDupModel = _.cloneDeep(self);
+            // var Collection = require('../../collection');
+            //
+            // console.log('connections', self.connections);
+            var colDupModel = self.constructor.extend({
```

Comment:
@particlebanana I am having some problems with this line: `self.constructor.extend`.
I am trying to `extend` (or `clone`) the existing Waterline Collection and change it's `identity` and add `baseIdentity` (identity of base Collection, without tenant suffix).

It works for my Unit Tests, however I am now using https://github.com/balderdashy/sails-mongo and it says that `self.constructor` is a `[Function: bound]` and that errors saying there is no method `extend`. 

```javascript
/Users/glavin/Documents/Streamlyne/api-server/node_modules/sails-mongo/node_modules/mongodb/lib/mongodb/mongo_client.js:475
          throw err
                ^
TypeError: Object function bound() {
        // `Function#bind` spec
        // http://es5.github.io/#x15.3.4.5
        if (partialArgs) {
          // avoid `arguments` object deoptimizations by using `slice` instead
          // of `Array.prototype.slice.call` and not assigning `arguments` to a
          // variable as a ternary expression
          var args = slice(partialArgs);
          push.apply(args, arguments);
        }
        // mimic the constructor's `return` behavior
        // http://es5.github.io/#x13.2.2
        if (this instanceof bound) {
          // ensure `new bound` is an instance of `func`
          var thisBinding = baseCreate(func.prototype),
              result = func.apply(thisBinding, args || arguments);
          return isObject(result) ? result : thisBinding;
        }
        return func.apply(thisArg, args || arguments);
      } has no method 'extend'
  at createTenantCollection (/Users/glavin/Documents/Project Dev/sails/node_modules/waterline/lib/waterline/query/dql/tenant.js:97:48)
  at /Users/glavin/Documents/Project Dev/sails/node_modules/waterline/lib/waterline/query/dql/tenant.js:171:24
  at /Users/glavin/Documents/Streamlyne/api-server/node_modules/sails-mongo/lib/adapter.js:110:9
  at /Users/glavin/Documents/Streamlyne/api-server/node_modules/sails-mongo/lib/connection.js:31:5
  at /Users/glavin/Documents/Streamlyne/api-server/node_modules/sails-mongo/node_modules/mongodb/lib/mongodb/mongo_client.js:472:11
  at process._tickDomainCallback (node.js:463:13)
```

Any help would be greatly appreciated!